titus_isolate/event/kubernetes_opportunistic_window_publisher.py

```diff
@@ -51,41 +60,68 @@ def __get_node(self) -> V1Node:
         log.debug('node: %s', node)
         return node
 
+    def __watch(self):
+        label_selector = "{}={}".format(OPPORTUNISTIC_RESOURCE_NODE_NAME_LABEL_KEY,
+                                        self.__node_name)
+        while True:
+            log.info("Starting opportunistic resource watch...")
+            try:
+                stream = watch.Watch().stream(
+                    self.__custom_api.list_cluster_custom_object,
+                    "titus.netflix.com",
+                    "v1",
+                    "opportunistic-resources",
+                    label_selector=label_selector)
+
+                for event in stream:
+                    log.info("Event: %s", event)
+                    event_type = event['type']
+                    event_metadata_name = event['object']['metadata']['name']
+
+                    with self.__lock:
+                        if event_type == 'ADDED':
+                            self.__opportunistic_resources[event_metadata_name] = event
+                        elif event_type == 'DELETED':
+                            self.__opportunistic_resources.pop(event_metadata_name, None)
+
+            except:
+                log.exception("Watch of opportunistic resources failed")
+
     def is_window_active(self) -> bool:
-        oppo_list = self.__get_scoped_opportunistic_resources()
-        log.debug('is active: oppo list: %s', json.dumps(oppo_list))
-        for item in oppo_list['items']:
-            log.debug('checking for window: %s', json.dumps(item))
-            now = datetime.utcnow()
-            if now < self.__get_timestamp(item['spec']['window']['end']):
-                return True
-        return False
+        with self.__lock:
+            log.debug('is active: oppo list: %s', json.dumps(self.__opportunistic_resources))
+            for item in self.__opportunistic_resources.values():
+                log.debug('checking for window: %s', json.dumps(item))
+                now = datetime.utcnow()
+                if now < self.__get_timestamp(item['object']['spec']['window']['end']):
```

Comment:
Hehe. This feels like playing with fire here, because utcnow has no TZ set. I looked at the get_timestamp function and it also has no TZ if it is an int, but will have a tz if it is a string. This is fine if both the api server and this code are in UTC.

I don't think we need to change anything right now, but this could be a source of tz bugs.

Reply:
This timestamp is read and written by this component.

Reply:
Oh, except when it's actually being used for something on the control plane.  Yes all Titus components run in UTC, but we may want to do something here.  I'll leave it be for now.